src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs

```diff
@@ -2761,6 +2761,110 @@ .maxstack  3
+
+dynamic hole = 1;
+Console.WriteLine(" + expression + @");
+";
+
+            var interpolatedStringBuilder = GetInterpolatedStringHandlerDefinition(includeSpanOverloads: false, useDefaultParameters: false, useBoolReturns: false);
+
+            var verifier = CompileAndVerifyWithCSharp(new[] { source, interpolatedStringBuilder }, expectedOutput: @"base1");
+
+            verifier.VerifyIL("<top-level-statements-entry-point>", expression.Contains('+')
+? @"
+{
+  // Code size       34 (0x22)
+  .maxstack  3
+  .locals init (object V_0) //hole
+  IL_0000:  ldc.i4.1
+  IL_0001:  box        ""int""
+  IL_0006:  stloc.0
+  IL_0007:  ldstr      ""base""
+  IL_000c:  ldstr      ""{0}""
+  IL_0011:  ldloc.0
+  IL_0012:  call       ""string string.Format(string, object)""
+  IL_0017:  call       ""string string.Concat(string, string)""
+  IL_001c:  call       ""void System.Console.WriteLine(string)""
+  IL_0021:  ret
+}
+"
+: @"
+{
+  // Code size       24 (0x18)
+  .maxstack  2
+  .locals init (object V_0) //hole
+  IL_0000:  ldc.i4.1
+  IL_0001:  box        ""int""
+  IL_0006:  stloc.0
+  IL_0007:  ldstr      ""base{0}""
+  IL_000c:  ldloc.0
+  IL_000d:  call       ""string string.Format(string, object)""
+  IL_0012:  call       ""void System.Console.WriteLine(string)""
+  IL_0017:  ret
+}
+");
+        }
+
+        [Theory, WorkItem(55609, "https://github.com/dotnet/roslyn/issues/55609")]
+        [InlineData(@"$""{hole}base""")]
+        [InlineData(@"$""{hole}"" + $""base""")]
+        public void DynamicInHoles_UsesFormat2(string expression)
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+dynamic hole = 1;
+Console.WriteLine(" + expression + @");
+";
+
+            var interpolatedStringBuilder = GetInterpolatedStringHandlerDefinition(includeSpanOverloads: false, useDefaultParameters: false, useBoolReturns: false);
+
+            var verifier = CompileAndVerifyWithCSharp(new[] { source, interpolatedStringBuilder }, expectedOutput: @"1base");
+
+            verifier.VerifyIL("<top-level-statements-entry-point>", expression.Contains('+')
+? @"
+{
+  // Code size       34 (0x22)
+  .maxstack  2
+  .locals init (object V_0) //hole
+  IL_0000:  ldc.i4.1
+  IL_0001:  box        ""int""
+  IL_0006:  stloc.0
+  IL_0007:  ldstr      ""{0}""
+  IL_000c:  ldloc.0
+  IL_000d:  call       ""string string.Format(string, object)""
+  IL_0012:  ldstr      ""base""
+  IL_0017:  call       ""string string.Concat(string, string)""
+  IL_001c:  call       ""void System.Console.WriteLine(string)""
+  IL_0021:  ret
+}
+"
+: @"
+{
+  // Code size       24 (0x18)
+  .maxstack  2
+  .locals init (object V_0) //hole
+  IL_0000:  ldc.i4.1
+  IL_0001:  box        ""int""
+  IL_0006:  stloc.0
+  IL_0007:  ldstr      ""{0}base""
+  IL_000c:  ldloc.0
+  IL_000d:  call       ""string string.Format(string, object)""
+  IL_0012:  call       ""void System.Console.WriteLine(string)""
+  IL_0017:  ret
+}
+");
+        }
```

Comment:
If I'm tracking properly, there's a few interesting scenarios: 
1. conversion to string (`BindUnconvertedInterpolatedStringToString`), 
2. conversion to default handler (`TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler`) and 
3. conversion to custom handler. 
4. We also have new permutations for binary addition expressions (`RebindSimpleBinaryOperatorAsConverted` calls `TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler`).

Are both new tests focusing on the `BindUnconvertedInterpolatedStringToString` change only?
Would it be good to cover the other scenarios as well?

Reply:
Scenarios 1, 2, and 4 are the same scenario, and all are covered by `DynamicInHoles_UsesFormat` and `DynamicInHoles_UsesFormat2`. Scenario 3 isn't really _interesting_, because it's not new: as Jared pointed out in the bug, we don't warn about dynamic/ref struct mixing in general. `RefStructHandler_DynamicInHole` covers this, but it verifies that we don't actually do anything special and the code will crash at runtime.